src/backend/vulkan/src/lib.rs

```diff
@@ -472,31 +436,27 @@ impl core::Adapter<Backend> for Adapter {
                 type_flags |= memory::LAZILY_ALLOCATED;
             }
 
-            core::MemoryType {
+            hal::MemoryType {
                 id: i,
                 properties: type_flags,
                 heap_index: mem.heap_index as usize,
             }
         }).collect();
 
-        let device_arc = device.raw.clone();
-        core::Gpu {
+        hal::Gpu {
             device,
-            general_queues: collect_queues(queue_descs, &device_arc, QueueType::General),
-            graphics_queues: collect_queues(queue_descs, &device_arc, QueueType::Graphics),
-            compute_queues: collect_queues(queue_descs, &device_arc, QueueType::Compute),
-            transfer_queues: collect_queues(queue_descs, &device_arc, QueueType::Transfer),
+            queue_families,
             memory_types,
             memory_heaps,
         }
     }
 
-    fn info(&self) -> &core::AdapterInfo {
+    fn info(&self) -> &hal::AdapterInfo {
         &self.info
     }
 
-    fn queue_families(&self) -> &[(QueueFamily, QueueType)] {
-        &self.queue_families
+    fn list_queue_families(&mut self) -> Vec<ProtoQueueFamily> {
+        mem::replace(&mut self.queue_families, Vec::new())
```

Comment:
This turned into some sort of 'one-shot' function. Do we want to preserve this or rather clone here?

Reply:
The idea is to have unique `ProtoQueueFamily` instances: user gets a list of them, passes some of them back when calling `open`, drops the other. The only non-fitting piece here is `create_command_pool`, which has to accept `&ProtoQueueFamily`, thus allowing the user to pass one of the prototypes not actually involved in the `Adapter::open`. Not sure how to fix that :( suggestions welcome!